src/mpid/ch4/shm/posix/eager/iqueue/iqueue_recv.h

```diff
@@ -0,0 +1,148 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+#ifndef POSIX_EAGER_IQUEUE_RECV_H_INCLUDED
+#define POSIX_EAGER_IQUEUE_RECV_H_INCLUDED
+
+#include "iqueue_impl.h"
+
+/* Work backward through the queue to find the first cell and put that one in the transport's
+ * first_cell position. */
+static void MPIDI_POSIX_EAGER_IQUEUE_enqueue_cell(MPIDI_POSIX_EAGER_IQUEUE_transport_t * transport,
+                                                  MPIDI_POSIX_EAGER_IQUEUE_cell_t * cell)
+{
+    /* Ingress queue of cells is inverted. Thus, we need to re-invert it. */
+    if (cell->prev) {
+        MPIDI_POSIX_EAGER_IQUEUE_cell_t *prev =
+            MPIDI_POSIX_EAGER_IQUEUE_GET_CELL(transport, cell->prev);
+        MPIDI_POSIX_EAGER_IQUEUE_enqueue_cell(transport, prev);
+        cell->prev = 0;
+    }
+    cell->next = NULL;
+    if (transport->last_cell) {
+        transport->last_cell->next = cell;
+    } else {
+        transport->first_cell = cell;
+    }
+    transport->last_cell = cell;
+}
+
+static inline MPIDI_POSIX_EAGER_IQUEUE_cell_t
+    * MPIDI_POSIX_EAGER_IQUEUE_receive_cell(MPIDI_POSIX_EAGER_IQUEUE_transport_t * transport)
+{
+    MPIDI_POSIX_EAGER_IQUEUE_cell_t *cell = NULL;
+    MPIDI_POSIX_EAGER_IQUEUE_terminal_t *terminal;
+
+    /* Update the first_cell pointer in the queue if we pull out the first cell */
+    if (transport->first_cell) {
+        cell = transport->first_cell;
+        if (cell->next == NULL) {
+            transport->first_cell = NULL;
+            transport->last_cell = NULL;
+        } else {
+            transport->first_cell = cell->next;
+        }
+    } else {
+        /* If first_cell wasn't set, grab the next cell from the appropriate terminal */
+        terminal = &transport->terminals[transport->local_rank];
+
+        if (terminal->head) {
+            uintptr_t head =
+                (uintptr_t) (unsigned int *) OPA_swap_ptr((OPA_ptr_t *) & terminal->head, NULL);
+
+            cell = MPIDI_POSIX_EAGER_IQUEUE_GET_CELL(transport, head);
+
+            /* If the head of the terminal has the prev pointer set, enqueue the current cell and
+             * use that one instead. */
+            if (cell->prev) {
+                MPIDI_POSIX_EAGER_IQUEUE_enqueue_cell(transport, cell);
+
+                cell = transport->first_cell;
+                MPIR_Assert(cell != NULL);
+                MPIR_Assert(cell->next != NULL);
+                transport->first_cell = cell->next;
+            }
+        }
+    }
+
```

Comment:
I don't understand the logic of this function. Can you please explain why we need `first_cell` and `last_cell` at transport ? 

Besides, casting `OPA_ptr_t` type seems unsafe.

Reply:
I believe they are used to tell the boundaries of a message that's larger than a single cell. At least, that's what the documentation I came up with says.

Is that true @SergeyGubanov?

Reply:
> Besides, casting OPA_ptr_t type seems unsafe.

This will need the same solution as https://github.com/pmodels/mpich/pull/3199#discussion_r224264249

Reply:
The terminal->head is a stack (inverted-ordered queue) of incomimg cells (public scope of visibility, visible to all ranks by shared memory).

The first_cell and last_cell is the true-ordered incoming queue (private scope of visibility, visible to the process which receive the cells).